### ISSUE TITLE
Incorrect weighting for Table 2 "Weighted by PITF" predictions?

### ISSUE THREAD

**Issue by andybega** [andybega/Blair-Sambanis-replication#3]
The "Weighted by PITF" model is a combination of the Escalation model predictions multiplied by the PITF model predictions. From the paper:

> The second (col- umn 3) uses PITF predicted probabilities to weight the results of the escalation model, ensuring that high-risk and low-risk countries that happen to take similar values on ICEWS-based predictors are nonetheless assigned different predicted probabilities in most months.

I think the weighting is incorrect. The test set predictions are weighted using the training data PITF model predictions. 

In `1mo_run_escalation_weighted_PITF.R`:

```r
            weight <- as.numeric(train$pred_prob_plus1)
            length(weight) <- length(prediction_escalation_inc_civil_ns)
            
            prediction = prediction_escalation_inc_civil_ns*weight
            prediction[is.na(escalation_test_frame_civil_ns$test_DV_civil_ns)] <- NA
```

`train$pred_prob_plus1` references the training data, set up on lines 43 and on in `+master.R`:

```r 
# Load data

data <- read.dta13("data/1mo_data.dta")

# Define training and testing sets for base specification

train_period = mean(data$period[which(data$month==12 & data$year==2007)])
end_period = mean(data$period[which(data$month==12 & data$year==2015)])
train <- data[data$period<=train_period,]
test <- data[data$period>train_period & data$period<=end_period,]
```

`prediction_escalation_inc_civil_ns` is created in `1mo_run_escalation.R` line 205:

```r
prediction_escalation_inc_civil_ns <- prediction
```

where `prediction` is in that same script higher up; just the base specification escalation predictions for the test set:

```r
prediction <- as.numeric(predict(train_model, newdata = data.matrix(escalation_test_frame_civil_ns), type="response"))
prediction[is.na(escalation_test_frame_civil_ns$test_DV_civil_ns)] <- NA
```

Check this. What are the object lengths? And if this is correct, then what is the AUC-ROC with the train/test PITF predictions? For both 1 and 6-month. This kind of stuff would be good to document.

**Comment by andybega** [comment]
> Q: Are we sure that B&S didn't intend to combine the PITF training data weights with test data weights?

I think so. I'm pretty confident that if you try to pursue this idea you could find some sort of way to make it work in a way that makes sense, but even if this was the intention and I am wrong to presume that it instead was to combine the PITF and Escalation test predictions, what the actual code does doesn't make any sense once you start digging more. 

To recap:

The crucial bit of code is in `1_mo_run_escalation_weighted_PITF.R`:

```r
weight <- as.numeric(train$pred_prob_plus1)
length(weight) <- length(prediction_escalation_inc_civil_ns)
```

`train` includes the ID variables already, what about the prediction vector?

This is created in `1_mo_run_escalation.R` lines 95-96:

```r
prediction <- as.numeric(predict(train_model, newdata = data.matrix(escalation_test_frame_civil_ns), type="response"))
prediction[is.na(escalation_test_frame_civil_ns$test_DV_civil_ns)] <- NA
```

(`prediction` is copied to `prediction_escalation_...` on line 205)

`escalation_test_frame_civil_ns` is assigned (using `assign()`) during a loop in `1mo_define_models.R`, see line 2532:

```r
  assign(paste0(model, "_train_frame_civil_ns"), data.frame(train_DV_civil_ns, eval(parse(text=paste0(model, "_train_RHS")))))
  assign(paste0(model, "_test_frame_civil_ns"), data.frame(test_DV_civil_ns, eval(parse(text=paste0(model, "_test_RHS")))))
```

None of those bits have ID vars. Where can we get ID vars? This is how long, i.e. how many country-month cases, this prediction vector and the associated outcome vector and data frame represent:
```
> length(test_DV_civil_ns)
[1] 15744
```

This matches the `test` data frame in `+master.R`'s `test`, so we can pull ID vars from there.

```
> nrow(test)
[1] 15744
```

Ok, so we have ID vars now. What happens when we multiply the vectors like above? What cases are we actually multiplying? 

```r
df1 <- train %>%
  select(country_name, year, month) %>%
  rename(country_train = country_name, year_train = year, month_train = month)
df2 <- test %>% 
  select(country_name, year, month) %>%
  rename(country_test = country_name, year_test = year, month_test = month) %>%
  # hack equivalent to what is happening above
  slice(1:nrow(df1))
foo <- cbind(df1, df2)
# reset row indices
rownames(foo) <- NULL
```

The data frame `foo` now maps training set cases from which we are taking PITF model predictions to the test set cases that contain the Escalation model predictions we are weighing. Now, `train` spans 7 years while `test` spans 8 years. Things are misaligned. So, we get weird things like:

Because we are extending the original N=13776 weights vector with 1968 NAs to reach N=15744, which is how many test set cases we have, the last 1968 cases in the test set essentially drop out (something times NA is NA). This means that all countries after Togo completely drop out of the test data.

```r
> tail(foo$country_test, 1)
[1] "Togo"
> unique(foo$country_train)[unique(foo$country_train) > "Togo"]
 [1] "United Arab Emirates" "United Kingdom"       "Turkmenistan"        
 [4] "Trinidad and Tobago"  "Tunisia"              "Turkey"              
 [7] "Uganda"               "Ukraine"              "Uruguay"             
[10] "United States"        "Uzbekistan"           "Venezuela"           
[13] "Vietnam"              "Yemen"                "Zambia"              
[16] "Zimbabwe"   
```

Because of the misalignment induced by matching 7 vs 8 year long sequences of countries, we get weird things like this weighting of the Escalation predictions: 

```r
> foo %>% filter(country_test=="Syria") %>% group_by(country_train, country_test) %>% summarize(year_train = paste0(range(year_train), collapse = " - "), year_test = paste0(range(year_test), collapse = " - "))
`summarise()` regrouping output by 'country_train' (override with `.groups` argument)
# A tibble: 2 x 4
# Groups:   country_train [2]
  country_train country_test year_train  year_test  
  <chr>         <chr>        <chr>       <chr>      
1 South Africa  Syria        2002 - 2007 2008 - 2013
2 Zambia        Syria        2001 - 2002 2014 - 2015
```

The "Weighted by PITF" model predictions for Syria from 2008 to 2015 are the Escalation test predictions for that time period weighted by the 2002 to 2007 PITF predictions for South Africa and the 2001 to 2002 PITF predictions for Zambia. 

That makes no sense at all.
